include/dlaf/permutations/general/impl.h

```diff
@@ -152,4 +157,345 @@ void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end, Matrix<con
                                                false>(dlaf::internal::Policy<B>(), std::move(permute_fn),
                                                       std::move(sender)));
 }
+
+template <class T, Device D>
+auto whenAllReadWriteTilesArray(LocalTileIndex begin, LocalTileIndex end, Matrix<T, D>& matrix) {
+  LocalTileSize sz{end.row() - begin.row() + 1, end.col() - begin.col() + 1};
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(ut::collectReadWriteTiles(begin, sz, matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadWriteTilesArray(Matrix<T, D>& matrix) {
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(
+      ut::collectReadWriteTiles(LocalTileIndex(0, 0), matrix.distribution().localNrTiles(), matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadOnlyTilesArray(LocalTileIndex begin, LocalTileIndex end, Matrix<const T, D>& matrix) {
+  LocalTileSize sz{end.row() - begin.row() + 1, end.col() - begin.col() + 1};
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(ut::collectReadTiles(begin, sz, matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadOnlyTilesArray(Matrix<const T, D>& matrix) {
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(
+      ut::collectReadTiles(LocalTileIndex(0, 0), matrix.distribution().localNrTiles(), matrix));
+}
+
+// No data is sent or received but the processes participates in the collective call
+template <class T>
+void all2allEmptyData(const comm::Communicator& comm) {
+  namespace ex = pika::execution::experimental;
+
+  auto all2all_f = [comm](const std::vector<int>& arr, MPI_Request* req) {
+    MPI_Datatype dtype = dlaf::comm::mpi_datatype<std::remove_pointer_t<T>>::type;
+    // Avoid buffer aliasing error reported when send_ptr and recv_ptr are the same.
+    int irrelevant = 42;
```

Comment:
Can you explain this in a bit more detail?

Reply:
I get a MPI error if I try to put `nullptr` where `&irrelevant` is placed with `mpich@3.2.1`. That is because the `send_ptr` and the `recv_ptr` must not alias. Technically, I think this is a minor MPI bug as I would expect the aliasing check to only happen if neither of the pointers is `nullptr`...

Reply:
I see now, thanks for the explanation!